File: apps/web/lib/db/queries/public.ts
import "server-only";

import { query, queryOne } from "@/lib/db/client";

/**
 * Read queries backing the public API.
 *
 * These are transliterations of the SQLAlchemy queries in
 * `services/api/app/routers/public.py`. Both implementations run side by side until
 * the Python routes are retired, and `api_baseline.py compare` proves they agree.
 *
 * Parameterized only — no string interpolation anywhere, same rule as the Python side.
 */

/* -------------------------------------------------------------------------- */
/* Row shapes                                                                  */
/* -------------------------------------------------------------------------- */

interface CategoryRow {
  slug: string;
  name: string;
  description: string | null;
  accent_token: string;
  is_commercial: boolean;
}

interface ArticleSummaryRow {
  public_id: string;
  slug: string;
  headline: string;
  dek: string;
  article_type: string;
  published_at: Date | null;
  updated_at_public: Date | null;
  first_published_at: Date | null;
  reading_time_seconds: number;
  is_sponsored: boolean;
  category_slug: string;
  category_name: string;
  category_description: string | null;
  category_accent_token: string;
  media_public_id: string | null;
  media_storage_key: string | null;
  media_width: number | null;
  media_height: number | null;
  media_alt_text: string | null;
  media_caption: string | null;
  media_credit: string | null;
  media_blurhash: string | null;
  media_rights_status: string | null;
  media_is_ai_generated: boolean | null;
  media_ai_disclosure_text: string | null;
}

interface ArticleDetailRow extends ArticleSummaryRow {
  body_blocks: unknown[];
  key_facts: string[];
  byline: string;
  structured_data: Record<string, unknown>;
  seo_title: string;
  meta_description: string;
  og_title: string;
  og_description: string;
  canonical_url: string | null;
  noindex: boolean;
  disclosure_text: string | null;
}

/* -------------------------------------------------------------------------- */
/* Serialization — mirrors _serialize_summary() in public.py                    */
/* -------------------------------------------------------------------------- */

/**
 * The public URL path, derived not stored — identical to `Article.path` in Python.
 * Month and day are zero-padded; an unpublished article has no public path.
 */
function articlePath(row: ArticleSummaryRow): string {
  if (row.first_published_at === null) {
    return `/preview/${row.public_id}`;
  }
  const d = row.first_published_at;
  const yyyy = d.getUTCFullYear();
  const mm = String(d.getUTCMonth() + 1).padStart(2, "0");
  const dd = String(d.getUTCDate()).padStart(2, "0");
  return `/${row.category_slug}/${yyyy}/${mm}/${dd}/${row.slug}`;
}

/**
 * Python emits `datetime.isoformat()`, which yields `+00:00` for a UTC-aware value —
 * NOT JavaScript's default `Z`. Matching this exactly matters: the baseline compares
 * the string.
 */
function isoUtc(value: Date | null): string | null {
  if (value === null) return null;
  return value.toISOString().replace(/\.(\d{3})Z$/, ".$1000+00:00");
}

function serializeSummary(row: ArticleSummaryRow): Record<string, unknown> {
  return {
    id: row.public_id,
    slug: row.slug,
    path: articlePath(row),
    headline: row.headline,
    dek: row.dek,
    articleType: row.article_type,
    category: {
      slug: row.category_slug,
      name: row.category_name,
      description: row.category_description,
      accentToken: row.category_accent_token,
    },
    publishedAt: isoUtc(row.published_at),
    updatedAt: isoUtc(row.updated_at_public),
    readingTimeSeconds: row.reading_time_seconds,
    isSponsored: row.is_sponsored,
    heroImage:
      row.media_public_id === null
        ? null
        : {
            id: row.media_public_id,
            url: `/media/${row.media_storage_key}`,
            width: row.media_width,
            height: row.media_height,
            altText: row.media_alt_text,
            caption: row.media_caption,
            credit: row.media_credit,
            blurhash: row.media_blurhash,
            rightsStatus: row.media_rights_status,
            isAiGenerated: row.media_is_ai_generated,
            aiDisclosure: row.media_ai_disclosure_text,
          },
  };
}

/* -------------------------------------------------------------------------- */
/* SQL                                                                         */
/* -------------------------------------------------------------------------- */

// Only published, non-deleted articles are ever visible publicly.
const PUBLISHED_SELECT = `
  SELECT
    a.public_id::text        AS public_id,
    a.slug,
    a.headline,
    a.dek,
    a.article_type,
    a.published_at,
    a.updated_at_public,
    a.first_published_at,
    a.reading_time_seconds,
    a.is_sponsored,
    c.slug                   AS category_slug,
    c.name                   AS category_name,
    c.description            AS category_description,
    c.accent_token           AS category_accent_token,
    m.public_id::text        AS media_public_id,
    m.storage_key            AS media_storage_key,
    m.width                  AS media_width,
    m.height                 AS media_height,
    m.alt_text               AS media_alt_text,
    m.caption                AS media_caption,
    m.credit                 AS media_credit,
    m.blurhash               AS media_blurhash,
    m.rights_status          AS media_rights_status,
    m.is_ai_generated        AS media_is_ai_generated,
    m.ai_disclosure_text     AS media_ai_disclosure_text
  FROM articles a
  JOIN categories c    ON c.id = a.category_id
  LEFT JOIN media_assets m ON m.id = a.hero_media_id
  WHERE a.status = 'published' AND a.deleted_at IS NULL
`;

export async function listCategories(): Promise<Record<string, unknown>[]> {
  const rows = await query<CategoryRow>(
    `SELECT slug, name, description, accent_token, is_commercial
       FROM categories
      WHERE is_active = true
      ORDER BY sort_order`,
  );
  return rows.map((row) => ({
    slug: row.slug,
    name: row.name,
    description: row.description,
    accentToken: row.accent_token,
    isCommercial: row.is_commercial,
  }));
}

export interface ArticleListResult {
  items: Record<string, unknown>[];
  page: number;
  pageSize: number;
  hasMore: boolean;
  total: number;
}

export async function listArticles(options: {
  category: string | null;
  page: number;
  pageSize: number;
}): Promise<ArticleListResult> {
  const { category, page, pageSize } = options;
  const offset = (page - 1) * pageSize;

  // Fetch one extra row to determine hasMore without a second COUNT query. `total` is
  // therefore an ESTIMATE, matching the Python contract exactly — a true count here
  // would be a behaviour change, not an improvement.
  const limit = pageSize + 1;

  const rows = category
    ? await query<ArticleSummaryRow>(
        `${PUBLISHED_SELECT} AND c.slug = $1
         ORDER BY a.published_at DESC
         OFFSET $2 LIMIT $3`,
        [category, offset, limit],
      )
    : await query<ArticleSummaryRow>(
        `${PUBLISHED_SELECT}
         ORDER BY a.published_at DESC
         OFFSET $1 LIMIT $2`,
        [offset, limit],
      );

  const hasMore = rows.length > pageSize;
  const items = rows.slice(0, pageSize);

  return {
    items: items.map(serializeSummary),
    page,
    pageSize,
    hasMore,
    total: offset + items.length + (hasMore ? 1 : 0),
  };
}

export async function listLatest(limit: number): Promise<Record<string, unknown>[]> {
  const rows = await query<ArticleSummaryRow>(
    `${PUBLISHED_SELECT} ORDER BY a.published_at DESC LIMIT $1`,
    [limit],
  );
  return rows.map(serializeSummary);
}

export async function getArticleBySlug(
  categorySlug: string,
  slug: string,
): Promise<Record<string, unknown> | null> {
  const row = await queryOne<ArticleDetailRow>(
    `SELECT
       a.body_blocks, a.key_facts, a.byline, a.structured_data,
       a.seo_title, a.meta_description, a.og_title, a.og_description,
       a.canonical_url, a.noindex, a.disclosure_text,
       sub.*
     FROM (${PUBLISHED_SELECT} AND c.slug = $1 AND a.slug = $2) sub
     JOIN articles a ON a.public_id::text = sub.public_id`,
    [categorySlug, slug],
  );

  if (row === null) return null;

  const [sources, corrections, tags] = await Promise.all([
    query<{ publisher: string; title: string; url: string; ref_type: string }>(
      `SELECT r.publisher, r.title, r.url, r.ref_type
         FROM article_source_refs r
         JOIN articles a ON a.id = r.article_id
        WHERE a.public_id::text = $1
        ORDER BY r.display_order`,
      [row.public_id],
    ),
    query<{ correction_type: string; summary: string; detail: string; issued_at: Date }>(
      `SELECT co.correction_type, co.summary, co.detail, co.issued_at
         FROM corrections co
         JOIN articles a ON a.id = co.article_id
        WHERE a.public_id::text = $1 AND co.is_public = true`,
      [row.public_id],
    ),
    query<{ slug: string; name: string }>(
      `SELECT t.slug, t.name
         FROM tags t
         JOIN article_tags at ON at.tag_id = t.id
         JOIN articles a ON a.id = at.article_id
        WHERE a.public_id::text = $1`,
      [row.public_id],
    ),
  ]);

  return {
    ...serializeSummary(row),
    body: row.body_blocks,
    keyFacts: row.key_facts,
    byline: row.byline,
    tags: tags.map((t) => ({ slug: t.slug, name: t.name })),
    sources: sources.map((s) => ({
      publisher: s.publisher,
      title: s.title,
      url: s.url,
      refType: s.ref_type,
    })),
    corrections: corrections.map((c) => ({
      type: c.correction_type,
      summary: c.summary,
      detail: c.detail,
      issuedAt: isoUtc(c.issued_at),
    })),
    seo: {
      // Python falls back to the headline/dek when the SEO field is blank.
      title: row.seo_title || row.headline,
      metaDescription: row.meta_description || row.dek,
      ogTitle: row.og_title || row.headline,
      ogDescription: row.og_description || row.dek,
      canonicalUrl: row.canonical_url || articlePath(row),
      noindex: row.noindex,
    },
    structuredData: row.structured_data,
    disclosure: row.disclosure_text,
  };
}

/** Exported for unit tests that pin the pagination contract. */
export const __testing = { articlePath, isoUtc, serializeSummary };

File: services/api/app/routers/public.py
"""Public read API.

Endpoints are *page-shaped*, not resource-shaped: each one returns exactly what a
page needs in a single round trip. Generic REST resources would push the web app into
N+1 API calls (ADR-0006).

Everything here is unauthenticated and cacheable. Nothing here can write.
"""

from __future__ import annotations

from datetime import UTC, datetime
from typing import Annotated, Any

from fastapi import APIRouter, HTTPException, Query, Response, status
from sqlalchemy import select
from sqlalchemy.orm import selectinload
from thedrop_database.enums import ArticleStatus
from thedrop_database.models import Article, Category

from app.deps import SessionDep

router = APIRouter(prefix="/api/v1/public", tags=["public"])

# Published content is immutable enough to cache hard at the edge; the publish task
# revalidates explicitly when something changes.
_CACHE_CONTROL = "public, max-age=60, stale-while-revalidate=300"


def _serialize_summary(article: Article) -> dict[str, Any]:
    hero = article.hero_media
    return {
        "id": str(article.public_id),
        "slug": article.slug,
        "path": article.path,
        "headline": article.headline,
        "dek": article.dek,
        "articleType": article.article_type,
        "category": {
            "slug": article.category.slug,
            "name": article.category.name,
            "description": article.category.description,
            "accentToken": article.category.accent_token,
        },
        "publishedAt": article.published_at.isoformat() if article.published_at else None,
        "updatedAt": (
            article.updated_at_public.isoformat() if article.updated_at_public else None
        ),
        "readingTimeSeconds": article.reading_time_seconds,
        "isSponsored": article.is_sponsored,
        "heroImage": (
            {
                "id": str(hero.public_id),
                "url": f"/media/{hero.storage_key}",
                "width": hero.width,
                "height": hero.height,
                "altText": hero.alt_text,
                "caption": hero.caption,
                "credit": hero.credit,
                "blurhash": hero.blurhash,
                "rightsStatus": hero.rights_status,
                "isAiGenerated": hero.is_ai_generated,
                "aiDisclosure": hero.ai_disclosure_text,
            }
            if hero
            else None
        ),
    }


def _published_query():
    return (
        select(Article)
        .options(selectinload(Article.category), selectinload(Article.hero_media))
        .where(Article.status == ArticleStatus.PUBLISHED, Article.deleted_at.is_(None))
        .order_by(Article.published_at.desc())
    )


@router.get("/categories")
def list_categories(db: SessionDep, response: Response) -> list[dict[str, Any]]:
    response.headers["Cache-Control"] = _CACHE_CONTROL
    categories = db.scalars(
        select(Category).where(Category.is_active.is_(True)).order_by(Category.sort_order)
    ).all()
    return [
        {
            "slug": c.slug,
            "name": c.name,
            "description": c.description,
            "accentToken": c.accent_token,
            "isCommercial": c.is_commercial,
        }
        for c in categories
    ]


@router.get("/articles")
def list_articles(
    db: SessionDep,
    response: Response,
    category: Annotated[str | None, Query(max_length=64)] = None,
    page: Annotated[int, Query(ge=1, le=500)] = 1,
    page_size: Annotated[int, Query(ge=1, le=50)] = 20,
) -> dict[str, Any]:
    response.headers["Cache-Control"] = _CACHE_CONTROL

    query = _published_query()
    if category:
        query = query.join(Category).where(Category.slug == category)

    offset = (page - 1) * page_size
    # Fetch one extra to determine hasMore without a second COUNT query.
    rows = db.scalars(query.offset(offset).limit(page_size + 1)).all()
    has_more = len(rows) > page_size
    items = rows[:page_size]

    return {
        "items": [_serialize_summary(a) for a in items],
        "page": page,
        "pageSize": page_size,
        "hasMore": has_more,
        "total": offset + len(items) + (1 if has_more else 0),
    }


@router.get("/articles/{category}/{year}/{month}/{day}/{slug}")
def get_article(
    db: SessionDep,
    response: Response,
    category: str,
    year: int,
    month: int,
    day: int,
    slug: str,
) -> dict[str, Any]:
    response.headers["Cache-Control"] = _CACHE_CONTROL

    article = db.scalar(
        _published_query()
        .options(
            selectinload(Article.source_refs),
            selectinload(Article.corrections),
            selectinload(Article.tags),
        )
        .join(Category)
        .where(Category.slug == category, Article.slug == slug)
    )
    if article is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Article not found")

    # The date is part of the canonical URL, so a mismatched date is a different URL
    # for the same content -- reject rather than serve duplicate paths.
    published = article.first_published_at
    if published is None or (published.year, published.month, published.day) != (year, month, day):
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Article not found")

    payload = _serialize_summary(article)
    payload.update(
        {
            "body": article.body_blocks,
            "keyFacts": article.key_facts,
            "byline": article.byline,
            "tags": [{"slug": t.slug, "name": t.name} for t in article.tags],
            "sources": [
                {
                    "publisher": ref.publisher,
                    "title": ref.title,
                    "url": ref.url,
                    "refType": ref.ref_type,
                }
                for ref in sorted(article.source_refs, key=lambda r: r.display_order)
            ],
            "corrections": [
                {
                    "type": c.correction_type,
                    "summary": c.summary,
                    "detail": c.detail,
                    "issuedAt": c.issued_at.isoformat(),
                }
                for c in article.corrections
                if c.is_public
            ],
            "seo": {
                "title": article.seo_title or article.headline,
                "metaDescription": article.meta_description or article.dek,
                "ogTitle": article.og_title or article.headline,
                "ogDescription": article.og_description or article.dek,
                "canonicalUrl": article.canonical_url or article.path,
                "noindex": article.noindex,
            },
            "structuredData": article.structured_data,
            "disclosure": article.disclosure_text,
        }
    )
    return payload


@router.get("/latest")
def latest(db: SessionDep, response: Response, limit: Annotated[int, Query(ge=1, le=50)] = 20):
    response.headers["Cache-Control"] = _CACHE_CONTROL
    rows = db.scalars(_published_query().limit(limit)).all()
    return {
        "items": [_serialize_summary(a) for a in rows],
        "generatedAt": datetime.now(UTC).isoformat(),
    }
